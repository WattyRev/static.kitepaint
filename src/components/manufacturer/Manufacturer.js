import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { H1, P, Spacer, PageLoader } from "../../theme";
import ProductsContainer from "../../containers/ProductsContainer";
import ManufacturerShowcase from "../create/ManufacturerShowcase";
import ProductShowcase from "../create/ProductShowcase";
import ErrorPage from "../ErrorPage";

/**
 * General styling for the create page
 */
const StyleWrapper = styled.div`
  margin: 0 auto;
  max-width: 1280px;

  .products-wrapper {
    display: flex;
    margin: 8px 0 24px;
    justify-content: space-around;
  }
  .heading {
    padding: 8px 16px;
  }
`;

/**
 * A coordinating component that builds the /create page.
 */
const Manufacturer = ({ match }) => (
  <ProductsContainer>
    {productsData => {
      const manufacturerName = decodeURIComponent(
        match.params.manufacturerName
      );
      if (productsData.props.isLoading) {
        return <PageLoader />;
      }
      if (
        !productsData.props.manufacturers.find(
          manufacturer => manufacturer.get("name") === manufacturerName
        )
      ) {
        return <ErrorPage />;
      }
      return (
        <StyleWrapper>
          <div className="heading">
            <H1>{manufacturerName}</H1>
            <P>Select a kite that you want to design</P>
          </div>
          {productsData.props.manufacturers.map(manufacturer => {
            const products =
              productsData.props.products[manufacturer.get("id")];
            if (!products || !products.length) {
              return null;
            }
            if (manufacturer.get("name") !== manufacturerName) {
              return null;
            }
            return (
              <Spacer key={manufacturer.get("id")} bottom="md">
                <ManufacturerShowcase manufacturer={manufacturer}>
                  <div className="products-wrapper">
                    {products.map(product => (
                      <ProductShowcase
                        product={product}
                        key={product.get("id")}
                      />
                    ))}
                  </div>
                </ManufacturerShowcase>
              </Spacer>
            );
          })}
        </StyleWrapper>
      );
    }}
  </ProductsContainer>
);

Manufacturer.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      manufacturerName: PropTypes.string
    })
  })
};

export default Manufacturer;
