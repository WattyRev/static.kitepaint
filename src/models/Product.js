import PropTypes from "prop-types";
import { createModel, computed } from "manikin-model";
import { getMockManufacturer } from "./Manufacturer";
import Status, { statusProp } from "./Status";

/**
 * A product is a kite that can be customized to create a design.
 */
const Product = createModel("Product", {
  id: null,
  embed: null,
  name: null,
  notes: null,
  manufacturer: null,
  status: null,
  url: null,
  colors: null,
  variations: null,

  // Raw manufacturer JSON
  json: computed(function() {
    return this.getProperties(
      "id",
      "embed",
      "name",
      "notes",
      "manufacturer",
      "status",
      "url",
      "colors",
      "variations"
    );
  })
});
Product.propTypes = {
  id: PropTypes.string.isRequired,
  embed: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  notes: PropTypes.arrayOf(PropTypes.string),
  manufacturer: PropTypes.string.isRequired,
  status: statusProp.isRequired,
  url: PropTypes.string,
  colors: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      color: PropTypes.string.isRequired
    })
  ),
  variations: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      name: PropTypes.string.isRequired,
      svg: PropTypes.string.isRequired
    })
  ),

  json: PropTypes.shape({
    id: PropTypes.string.isRequired,
    embed: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    notes: PropTypes.arrayOf(PropTypes.string),
    manufacturer: PropTypes.string.isRequired,
    status: statusProp.isRequired,
    url: PropTypes.string,
    colors: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        color: PropTypes.string.isRequired
      })
    ),
    variations: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        name: PropTypes.string.isRequired,
        svg: PropTypes.string.isRequired
      })
    )
  })
};

export default Product;

/**
 * A mock product used for testing
 * @type {Object}
 */
const getMockProduct = (overrides = {}) =>
  new Product(
    Object.assign(
      {
        id: "product-1",
        embed: "krazykites.com",
        name: "Krazy Kite",
        manufacturer: getMockManufacturer().get("id"),
        url: "http://krazykites.com/krazy-kite",
        colors: [
          {
            name: "red",
            color: "#ff0000"
          },
          {
            name: "black",
            color: "#000000"
          }
        ],
        variations: [
          {
            id: "1",
            name: "Standard",
            svg: `<div>Kool Kite<div data-id="1"></div></div>`
          }
        ],
        status: Status.PUBLIC
      },
      overrides
    )
  );

export { getMockProduct };
