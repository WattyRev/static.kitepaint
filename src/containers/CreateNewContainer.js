import { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { getProductById } from "../redux/modules/products";
import { getManufacturerByProductId } from "../redux/modules/manufacturers";
import { GET_PRODUCTS, GET_MANUFACTURERS } from "../redux/actions";
import Product from "../models/Product";
import Manufacturer from "../models/Manufacturer";

export const CreateNewContainer = ({
  product,
  manufacturer,
  onRequestProduct,
  onRequestManufacturer,
  children
}) => {
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    (async () => {
      try {
        await Promise.all([onRequestProduct(), onRequestManufacturer()]);
        setIsLoading(false);
      } catch (error) {
        setIsLoading(false);
      }
    })();
  }, []);

  return children({
    props: {
      isLoading,
      product,
      manufacturer
    }
  });
};

CreateNewContainer.propTypes = {
  /**
   * The ID of the product to get information about.
   */
  productId: PropTypes.string.isRequired,
  /**
   * The product. Provided by redux.
   */
  product: PropTypes.instanceOf(Product),
  /**
   * The manufactuer of the product. Provided by redux.
   */
  manufacturer: PropTypes.instanceOf(Manufacturer),
  /**
   * A function to request the product be fetched. Provided by redux.
   */
  onRequestProduct: PropTypes.func.isRequired,
  /**
   * A function to request the manufacturer of the product be fetched. Provided by redux.
   */
  onRequestManufacturer: PropTypes.func.isRequired,
  /**
   * A function that returns the content to be rendered.
   */
  children: PropTypes.func.isRequired
};

const mapStateToProps = (state, props) => ({
  product: getProductById(state, props.productId),
  manufacturer: getManufacturerByProductId(state, props.productId)
});

const mapDispatchToProps = {
  onRequestProduct: GET_PRODUCTS,
  onRequestManufacturer: GET_MANUFACTURERS
};

export default connect(mapStateToProps, mapDispatchToProps)(CreateNewContainer);
