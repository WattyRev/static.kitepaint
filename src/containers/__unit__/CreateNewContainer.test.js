import React from "react";
import { render, screen } from "@testing-library/react";
import { getMockProduct } from "../../models/Product";
import { getMockManufacturer } from "../../models/Manufacturer";
import { CreateNewContainer } from "../CreateNewContainer";

describe("CreateNewContainer", () => {
  let defaultProps;
  beforeEach(() => {
    defaultProps = {
      productId: "abc",
      onRequestProduct: jest.fn(),
      onRequestManufacturer: jest.fn()
    };
  });
  it("renders", () => {
    render(
      <CreateNewContainer {...defaultProps}>
        {() => <div data-testid="target">test</div>}
      </CreateNewContainer>
    );
    expect(screen.getByTestId("target")).toHaveTextContent("test");
  });
  it("requests for products to be fetched if no product was provided", () => {
    render(
      <CreateNewContainer {...defaultProps}>
        {() => <div>test</div>}
      </CreateNewContainer>
    );
    expect(defaultProps.onRequestProduct).toHaveBeenCalled();
  });
  it("requests for manufacturers to be fetched if no manufacturer was provided", () => {
    render(
      <CreateNewContainer {...defaultProps}>
        {() => <div>test</div>}
      </CreateNewContainer>
    );
    expect(defaultProps.onRequestManufacturer).toHaveBeenCalled();
  });
  it("displays the child render if the manufacturer and product are provided", () => {
    defaultProps.manufacturer = getMockManufacturer();
    defaultProps.product = getMockProduct();
    render(
      <CreateNewContainer {...defaultProps}>
        {data => (
          <div data-testid="content">
            {data.props.product.get("name")} by{" "}
            {data.props.manufacturer.get("name")}
          </div>
        )}
      </CreateNewContainer>
    );
    expect(screen.getByTestId("content")).toHaveTextContent(
      "Krazy Kite by Krazy Kites"
    );
  });
});
