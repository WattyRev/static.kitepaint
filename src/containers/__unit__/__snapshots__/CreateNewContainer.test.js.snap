// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNewContainer renders 1`] = `
<CreateNewContainer
  onRequestManufacturer={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onRequestProduct={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  productId="abc"
>
  <div>
    test
  </div>
</CreateNewContainer>
`;
